docs/doc/60-contributing/03-rfcs/20220602-async-insert.md

```diff
@@ -0,0 +1,136 @@
+
+After doing this, inserts into the same table will be prased and planned individual. The insert data will be convert to `DataBlock` and buffered. When some conditions are triggerd, the buffered `DataBlock` will be interpreted once which also is benifical for pipelines.
+
+# Design
+
+## Buffer Structure
+In order to buffer data of different tables. We need a `HashMap` to save datablocks of different insert plan. The settings of inserts also should be considered. But the settings of `QueryContext` take up big space which is unnecessary. Therefore, we save the changed settings. It's the settings which are differnt from default values of settings.
+
+**The key of HashMap:**
+
+```rust
+#[derive(Clone)]
+pub struct InsertKey {
+    plan: Arc<InsertPlan>,
+    // settings different with default settings
+    changed_settings: Settings,
+}
+```
+
+The plan is the first plan arrived at the bucket of hashmap. For example there comes two inserts:
+
+```rust
+insert into test values(1);
+insert into test select 1;
+```
+
+They are same plan because they have same schema and share one table. So the datablock of these two inserts will be put the same bucket. The plan of InsertKey is the plan `insert into test values(1);` which is insignificant. Because we only need the plan to construct `Interpreter`.
+
+**The Value of HashMap:**
+
+```rust
+pub struct InsertData {
+    entries: Vec<EntryPtr>,
+    data_size: u64,
+    first_update: Instant,
+    last_update: Instant,
+}
+```
+
+The `entries` is array of `Entry`:
+
+```rust
+pub struct Entry {
+    block: DataBlock,
+    finished: Arc<RwLock<bool>>,
+    notify: Arc<Notify>,
+}
+```
+
+Each insert corresponds to each `block`. The `finished` and `notify` is for query response. When server completed a batch of inserts, the entry will be finishd and notified.
+
+The `data_size` is the meomry size of `entries`. The `first_update` is the timestamp of the first insert into the bucket. The `last_update` is the timestamp of the last insert into the bucket. They all are used to trigger the batch inserts.
+
+## Execution
+
+Let's look at the `AsyncInsertQueue`:
+
+```rust
+type Queue = HashMap<InsertKey, InsertData>;
+type QueryIdToEntry = HashMap<String, EntryPtr>;
+
+pub struct AsyncInsertQueue {
+    pub session_mgr: Arc<RwLock<Option<Arc<SessionManager>>>>,
+    runtime: Arc<Runtime>,
+    max_data_size: u64,
+    busy_timeout: Duration,
+    stale_timeout: Duration,
+    queue: Arc<RwLock<Queue>>,
+    current_processing_insert: Arc<RwLock<QueryIdToEntry>>,
+}
+
+impl AsyncInsertQueue {
+    pub fn try_create() -> Self {}
+    pub async fn start(self: Arc<Self>) {}
+    pub async fn push(self: Arc<Self>, plan_node: Arc<InsertPlan>, ctx: Arc<QueryContext>) -> Result<()> {}
+    pub async fn wait_for_processing_insert(self: Arc<Self>, query_id: String, time_out: Duration,) -> Result<()> {}
```

Comment:
We'd better use `std::sync::mpsc::channel` to get insert result.

Reply:
Ok, I try it.